cli/internal/services/authentication/authentication.go

```diff
@@ -1,35 +1,87 @@
 package authentication
 
 import (
+	"cli/internal/services/utils"
 	"encoding/json"
+	"fmt"
+	"net/http"
 	"os"
 	"path/filepath"
+	"sync"
 )
 
-const (
-	fileName = "session"
-)
-
+// Authentication tokens.
 type Authentication struct {
 	SessionToken string `json:"access_token,omitempty"`
 	RefreshToken string `json:"refresh_token,omitempty"`
 }
 
+const (
+	// The name of the file where the session and refresh token are stored.
+	fileName = "session"
+)
+
+var (
+	// The Keycloak Configuration
+	// TODO: Read this from a config file or environment variables
+	keycloakConfig = KeycloakConfig{
+		URL: "http://localhost:8080",
+		Realm: "ihp-realm",
+		ClientId: "ihp-cli",
+	}
+
+	// The Embedded Server Configuration
+	// TODO: Read this from a config file or environment variables
+	embeddedServerConfig = utils.EmbeddedServerConfig{
+		Port: 3000,
+		CallbackPath: "/sso-cb",
+	}
+)
+
 // Starts the login process.
 // This will open the browser and redirect the user to keycloak.
 // After the user logged in, the session and refresh token will be saved.
 // TODO: If the user is already logged in, the login process will be skipped.
 func Login() error {
-	// Pretend we getting a session and refresh token from keycloak
-	sessionToken:= "sessionToken"
-	refreshToken:= "refreshToken"
-
-	tokens := Authentication{	
-		SessionToken: sessionToken,
-		RefreshToken: refreshToken,
+	// Open the browser and redirect the user to the internal server
+	if err := utils.OpenBrowser(keycloakConfig.getLoginURL("http://" + embeddedServerConfig.CallbackURL())); err != nil {
+		return fmt.Errorf("failed to open browser: %w", err)
 	}
 
-	return saveTokens(tokens);
+	// Wait for the embedded server to receive the callback
+	var asyncError error
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	// Start the embedded server
+	embeddedServerConfig.StartServer(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
a) This runs forever, right? Maybe a max time would be useful (although this may be tricky to implement)
b) Does this mean that the login is instantly considered failed if the embedded server endpoint is called? This may be something that we should keep in mind for the future

Reply:
Discussed